lib/rubygems/compiler.rb

```diff
@@ -47,6 +47,8 @@ def compile
       gemspec.files.push file
     end
 
+    gemspec.files = gemspec.files.select {|f| File.exists?("#{target_dir}/#{f}") } if @options[:purge]
```

Comment:
Originally I used ```select!``` but got some Travis failures for Rubies 1.8.7 (method didn't exist back then). Will try ```reject!```